File: library/src/main/java/com/panoramagl/PLObject.java
/*
 * PanoramaGL library
 * Version 0.2 beta
 * Copyright (c) 2010 Javier Baez <javbaezga@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.panoramagl;

import com.panoramagl.computation.PLMath;
import com.panoramagl.structs.PLPosition;
import com.panoramagl.structs.PLRange;
import com.panoramagl.structs.PLRotation;

public class PLObject extends PLObjectBase implements PLIObject
{
	/**member variables*/
	
	private boolean mIsXAxisEnabled, mIsYAxisEnabled, mIsZAxisEnabled;
	private PLPosition mPosition;
	private PLRange mXRange, mYRange, mZRange;
	
	private boolean mIsPitchEnabled, mIsYawEnabled, mIsRollEnabled, mIsReverseRotation, mIsYZAxisInverseRotation;
	private PLRotation mRotation;
	private PLRange mPitchRange, mYawRange, mRollRange, mTempRange;
	
	private float mAlpha, mDefaultAlpha;
	
	/**init methods*/
	
	public PLObject()
	{
		super();
	}
	
	@Override
	protected void initializeValues()
	{
		mXRange = PLRange.PLRangeMake(PLConstants.kFloatMinValue, PLConstants.kFloatMaxValue);
		mYRange = PLRange.PLRangeMake(PLConstants.kFloatMinValue, PLConstants.kFloatMaxValue);
		mZRange = PLRange.PLRangeMake(PLConstants.kFloatMinValue, PLConstants.kFloatMaxValue);
		
		mPitchRange = PLRange.PLRangeMake(PLConstants.kDefaultPitchMinRange, PLConstants.kDefaultPitchMaxRange);
		mYawRange = PLRange.PLRangeMake(PLConstants.kDefaultYawMinRange, PLConstants.kDefaultYawMaxRange);
		mRollRange = PLRange.PLRangeMake(PLConstants.kDefaultRollMinRange, PLConstants.kDefaultRollMaxRange);
		mTempRange = PLRange.PLRangeMake(0.0f, 0.0f);
		
		mIsXAxisEnabled = mIsYAxisEnabled = mIsZAxisEnabled = true;
		mIsPitchEnabled = mIsYawEnabled = mIsRollEnabled = true;
		
		mIsReverseRotation = false;
		
		mIsYZAxisInverseRotation = true;
		
		mPosition = PLPosition.PLPositionMake(0.0f, 0.0f, 0.0f);
		mRotation = PLRotation.PLRotationMake(0.0f, 0.0f, 0.0f);
		
		mAlpha = mDefaultAlpha = PLConstants.kDefaultAlpha;
	}
	
	/**reset methods*/
	
	@Override
	public void reset()
	{
		this.setRotation(0.0f, 0.0f, 0.0f);
		this.setInternalAlpha(mDefaultAlpha);
	}
	
	/**property methods*/
	
	@Override
	public PLPosition getPosition()
	{
		return mPosition;
	}
	
	@Override
	public void setPosition(PLPosition position)
	{
		if(position != null)
		{
			this.setX(position.x);
			this.setY(position.y);
			this.setZ(position.z);
		}
	}
	
	@Override
	public void setPosition(float x, float y, float z)
	{
		this.setX(x);
		this.setY(y);
		this.setZ(z);
	}
	
	@Override
	public float getX()
	{
		return mPosition.x;	
	}
	
	@Override
	public void setX(float x)
	{
		if(mIsXAxisEnabled)
			mPosition.x = PLMath.valueInRange(x, mXRange);
	}
	
	protected void setInternalX(float x)
	{
		mPosition.x = PLMath.valueInRange(x, mXRange);
	}
	
	@Override
	public float getY()
	{
		return mPosition.y;
	}
	
	@Override
	public void setY(float y)
	{
		if(mIsYAxisEnabled)
			mPosition.y = PLMath.valueInRange(y, mYRange);
	}
	
	protected void setInternalY(float y)
	{
		mPosition.y = PLMath.valueInRange(y, mYRange);
	}
	
	@Override
	public float getZ()
	{
		return mPosition.z;
	}
	
	@Override
	public void setZ(float z)
	{
		if(mIsZAxisEnabled)
			mPosition.z = PLMath.valueInRange(z, mZRange);
	}
	
	protected void setInternalZ(float z)
	{
		mPosition.z = PLMath.valueInRange(z, mZRange);
	}
	
	@Override
	public PLRotation getRotation()
	{
		return mRotation;
	}
	
	@Override
	public void setRotation(PLRotation rotation)
	{
		if(rotation != null)
		{
			this.setPitch(rotation.pitch);
			this.setYaw(rotation.yaw);
			this.setRoll(rotation.roll);
		}
	}
	
	@Override
	public void setRotation(float pitch, float yaw)
	{
		this.setPitch(pitch);
		this.setYaw(yaw);
	}
	
	@Override
	public void setRotation(float pitch, float yaw, float roll)
	{
		this.setPitch(pitch);
		this.setYaw(yaw);
		this.setRoll(roll);
	}
	
	@Override
	public float getPitch()
	{
		return mRotation.pitch;
	}
	
	@Override
	public void setPitch(float pitch)
	{
		if(mIsPitchEnabled)
			mRotation.pitch = this.getRotationAngleNormalized(pitch, mPitchRange);
	}
	
	protected void setInternalPitch(float pitch)
	{
		mRotation.pitch = this.getRotationAngleNormalized(pitch, mPitchRange);
	}
	
	@Override
	public float getYaw()
	{
		return mRotation.yaw;
	}
	
	@Override
	public void setYaw(float yaw)
	{
		if(mIsYawEnabled)
			mRotation.yaw = this.getRotationAngleNormalized(yaw, mYawRange);
	}
	
	protected void setInternalYaw(float yaw)
	{
		mRotation.yaw = this.getRotationAngleNormalized(yaw, mYawRange);
	}
	
	@Override
	public float getRoll()
	{
		return mRotation.roll;
	}
	
	@Override
	public void setRoll(float roll)
	{
		if(mIsRollEnabled)
			mRotation.roll = this.getRotationAngleNormalized(roll, mRollRange);
	}
	
	protected void setInternalRoll(float roll)
	{
		mRotation.roll = this.getRotationAngleNormalized(roll, mRollRange);
	}
	
	@Override
	public boolean isXAxisEnabled()
	{
		return mIsXAxisEnabled;
	}
	
	@Override
	public void setXAxisEnabled(boolean isXAxisEnabled)
	{
		mIsXAxisEnabled = isXAxisEnabled;
	}
	
	@Override
	public boolean isYAxisEnabled()
	{
		return mIsYAxisEnabled;
	}
	
	@Override
	public void setYAxisEnabled(boolean isYAxisEnabled)
	{
		mIsYAxisEnabled = isYAxisEnabled;
	}
	
	@Override
	public boolean isZAxisEnabled()
	{
		return mIsZAxisEnabled;
	}
	
	@Override
	public void setZAxisEnabled(boolean isZAxisEnabled)
	{
		mIsZAxisEnabled = isZAxisEnabled;
	}
	
	@Override
	public PLRange getXRange()
	{
		return mXRange;
	}
	
	@Override
	public void setXRange(PLRange xRange)
	{
		mXRange.setValues(xRange);
	}
	
	@Override
	public void setXRange(float min, float max)
	{
		mXRange.setValues(min, max);
	}
	
	protected void setInternalXRange(float min, float max)
	{
		mXRange.setValues(min, max);
	}
	
	@Override
	public float getXMin()
	{
		return mXRange.min;
	}
	
	@Override
	public void setXMin(float min)
	{
		mXRange.min = min;
	}
	
	@Override
	public float getXMax()
	{
		return mXRange.max;
	}
	
	@Override
	public void setXMax(float max)
	{
		mXRange.max = max;
	}
	
	@Override
	public PLRange getYRange()
	{
		return mYRange;
	}
	
	@Override
	public void setYRange(PLRange yRange)
	{
		mYRange.setValues(yRange);
	}
	
	@Override
	public void setYRange(float min, float max)
	{
		mYRange.setValues(min, max);
	}
	
	protected void setInternalYRange(float min, float max)
	{
		mYRange.setValues(min, max);
	}
	
	@Override
	public float getYMin()
	{
		return mYRange.min;
	}
	
	@Override
	public void setYMin(float min)
	{
		mYRange.min = min;
	}
	
	@Override
	public float getYMax()
	{
		return mYRange.max;
	}
	
	@Override
	public void setYMax(float max)
	{
		mYRange.max = max;
	}
	
	@Override
	public PLRange getZRange()
	{
		return mZRange;
	}
	
	@Override
	public void setZRange(PLRange zRange)
	{
		mZRange.setValues(zRange);
	}
	
	@Override
	public void setZRange(float min, float max)
	{
		mZRange.setValues(min, max);
	}
	
	protected void setInternalZRange(float min, float max)
	{
		mZRange.setValues(min, max);
	}
	
	@Override
	public float getZMin()
	{
		return mZRange.min;
	}
	
	@Override
	public void setZMin(float min)
	{
		mZRange.min = min;
	}
	
	@Override
	public float getZMax()
	{
		return mZRange.max;
	}
	
	@Override
	public void setZMax(float max)
	{
		mZRange.max = max;
	}
	
	@Override
	public boolean isPitchEnabled()
	{
		return mIsPitchEnabled;
	}
	
	@Override
	public void setPitchEnabled(boolean isPitchEnabled)
	{
		mIsPitchEnabled = isPitchEnabled;
	}
	
	@Override
	public boolean isYawEnabled()
	{
		return mIsYawEnabled;
	}
	
	@Override
	public void setYawEnabled(boolean isYawEnabled)
	{
		mIsYawEnabled = isYawEnabled;
	}
	
	@Override
	public boolean isRollEnabled()
	{
		return mIsRollEnabled;
	}
	
	@Override
	public void setRollEnabled(boolean isRollEnabled)
	{
		mIsRollEnabled = isRollEnabled;
	}
	
	@Override
	public boolean isReverseRotation()
	{
		return mIsReverseRotation;
	}
	
	@Override
	public void setReverseRotation(boolean isReverseRotation)
	{
		mIsReverseRotation = isReverseRotation;
	}
	
	@Override
	public boolean isYZAxisInverseRotation()
	{
		return mIsYZAxisInverseRotation;
	}
	
	@Override
	public void setYZAxisInverseRotation(boolean isYZAxisInverseRotation)
	{
		mIsYZAxisInverseRotation = isYZAxisInverseRotation;
	}
	
	@Override
	public PLRange getPitchRange()
	{
		return mPitchRange;
	}
	
	@Override
	public void setPitchRange(PLRange pitchRange)
	{
		mPitchRange.setValues(pitchRange);
	}
	
	@Override
	public void setPitchRange(float min, float max)
	{
		mPitchRange.setValues(min, max);
	}
	
	protected void setInternalPitchRange(float min, float max)
	{
		mPitchRange.setValues(min, max);
	}
	
	@Override
	public float getPitchMin()
	{
		return mPitchRange.min;
	}
	
	@Override
	public void setPitchMin(float min)
	{
		mPitchRange.min = min;
	}
	
	@Override
	public float getPitchMax()
	{
		return mPitchRange.max;
	}
	
	@Override
	public void setPitchMax(float max)
	{
		mPitchRange.max = max;
	}
	
	@Override
	public PLRange getYawRange()
	{
		return mYawRange;
	}
	
	@Override
	public void setYawRange(PLRange yawRange)
	{
		mYawRange.setValues(yawRange);
	}
	
	@Override
	public void setYawRange(float min, float max)
	{
		mYawRange.setValues(min, max);
	}
	
	protected void setInternalYawRange(float min, float max)
	{
		mYawRange.setValues(min, max);
	}
	
	@Override
	public float getYawMin()
	{
		return mYawRange.min;
	}
	
	@Override
	public void setYawMin(float min)
	{
		mYawRange.min = min;
	}
	
	@Override
	public float getYawMax()
	{
		return mYawRange.max;
	}
	
	@Override
	public void setYawMax(float max)
	{
		mYawRange.max = max;
	}
	
	@Override
	public PLRange getRollRange()
	{
		return mRollRange;
	}
	
	@Override
	public void setRollRange(PLRange rollRange)
	{
		mRollRange.setValues(rollRange);
	}
	
	@Override
	public void setRollRange(float min, float max)
	{
		mRollRange.setValues(min, max);
	}
	
	protected void setInternalRollRange(float min, float max)
	{
		mRollRange.setValues(min, max);
	}
	
	@Override
	public float getRollMin()
	{
		return mRollRange.min;
	}
	
	@Override
	public void setRollMin(float min)
	{
		mRollRange.min = min;
	}
	
	@Override
	public float getRollMax()
	{
		return mRollRange.max;
	}
	
	@Override
	public void setRollMax(float max)
	{
		mRollRange.max = max;
	}
	
	protected PLRange getTempRange()
	{
		return mTempRange;
	}
	
	@Override
	public float getAlpha()
	{
		return mAlpha;
	}
	
	@Override
	public void setAlpha(float alpha)
	{
		mAlpha = alpha;
	}
	
	protected void setInternalAlpha(float alpha)
	{
		mAlpha = alpha;
	}
	
	@Override
	public float getDefaultAlpha()
	{
		return mDefaultAlpha;
	}
	
	@Override
	public void setDefaultAlpha(float defaultAlpha)
	{
		mDefaultAlpha = defaultAlpha;
	}
	
	protected void setInternalDefaultAlpha(float defaultAlpha)
	{
		mDefaultAlpha = defaultAlpha;
	}
	
	/**normalize methods*/
	
	protected float getRotationAngleNormalized(float angle, PLRange range)
	{
		return PLMath.normalizeAngle(angle, mTempRange.setValues(-range.max, -range.min));
	}
	
	/**translate methods*/
	
	@Override
	public void translate(PLPosition position)
	{
		if(position != null)
		{
			this.setX(position.x);
			this.setY(position.y);
			this.setZ(position.z);
		}
	}
	
	@Override
	public void translate(float x, float y)
	{
		this.setX(x);
		this.setY(y);
	}
	
	@Override
	public void translate(float x, float y, float z)
	{
		this.setX(x);
		this.setY(y);
		this.setZ(z);
	}
	
	/**rotate methods*/
	
	@Override
	public void rotate(PLRotation rotation)
	{
		if(rotation != null)
		{
			this.setPitch(rotation.pitch);
			this.setYaw(rotation.yaw);
			this.setRoll(rotation.roll);
		}
	}
	
	@Override
	public void rotate(float pitch, float yaw)
	{
		this.setPitch(pitch);
		this.setYaw(yaw);
	}
	
	@Override
	public void rotate(float pitch, float yaw, float roll)
	{
		this.setPitch(pitch);
		this.setYaw(yaw);
		this.setRoll(roll);
	}
	
	/**clone methods*/
	
	@Override
	public boolean clonePropertiesOf(PLIObject object)
	{
		if(object != null)
		{
			this.setXAxisEnabled(object.isXAxisEnabled());
			this.setYAxisEnabled(object.isYAxisEnabled());
			this.setZAxisEnabled(object.isZAxisEnabled());
			
			this.setPitchEnabled(object.isPitchEnabled());
			this.setYawEnabled(object.isYawEnabled());
			this.setRollEnabled(object.isRollEnabled());
			
			this.setReverseRotation(object.isReverseRotation());
			
			this.setYZAxisInverseRotation(object.isYZAxisInverseRotation());
			
			this.setXRange(object.getXRange());
			this.setYRange(object.getYRange());
			this.setZRange(object.getZRange());
			
			this.setPitchRange(object.getPitchRange());
			this.setYawRange(object.getYawRange());
			this.setRollRange(object.getRollRange());
			
			this.setX(object.getX());
			this.setY(object.getY());
			this.setZ(object.getZ());
			
			this.setPitch(object.getPitch());
			this.setYaw(object.getYaw());
			this.setRoll(object.getRoll());
			
			this.setDefaultAlpha(object.getDefaultAlpha());
			this.setAlpha(object.getAlpha());
			
			return true;
		}
		return false;
	}
	
	/**dealloc methods*/
	
	@Override
	protected void finalize() throws Throwable
	{
		mPosition = null;
		mXRange = mYRange = mZRange = mTempRange = null;
		mRotation = null;
		mPitchRange = mYawRange = mRollRange = null;
		super.finalize();
	}
}

File: library/src/main/java/com/panoramagl/PLObject.kt
/*
 * PanoramaGL library
 * Version 0.2 beta
 * Copyright (c) 2010 Javier Baez <javbaezga@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.panoramagl

import com.panoramagl.computation.PLMath
import com.panoramagl.structs.PLPosition
import com.panoramagl.structs.PLRange
import com.panoramagl.structs.PLRange.Companion.PLRangeMake
import com.panoramagl.structs.PLRotation

open class PLObject : PLObjectBase(), PLIObject {
    override var isXAxisEnabled: Boolean = false
    override var isYAxisEnabled: Boolean = false
    override var isZAxisEnabled: Boolean = false
    private var mPosition: PLPosition = PLPosition()
    private var mXRange: PLRange? = null
    private var mYRange: PLRange? = null
    private var mZRange: PLRange? = null

    override var isPitchEnabled: Boolean = false
    override var isYawEnabled: Boolean = false
    override var isRollEnabled: Boolean = false
    override var isReverseRotation: Boolean = false
    override var isYZAxisInverseRotation: Boolean = false
    private var mRotation: PLRotation = PLRotation()
    private var mPitchRange: PLRange? = null
    private var mYawRange: PLRange? = null
    private var mRollRange: PLRange? = null
    protected var tempRange: PLRange? = null
        private set

    override var alpha: Float = 0f
    override var defaultAlpha: Float = 0f

    override fun initializeValues() {
        mXRange = PLRangeMake(PLConstants.kFloatMinValue, PLConstants.kFloatMaxValue)
        mYRange = PLRangeMake(PLConstants.kFloatMinValue, PLConstants.kFloatMaxValue)
        mZRange = PLRangeMake(PLConstants.kFloatMinValue, PLConstants.kFloatMaxValue)

        mPitchRange = PLRangeMake(PLConstants.kDefaultPitchMinRange, PLConstants.kDefaultPitchMaxRange)
        mYawRange = PLRangeMake(PLConstants.kDefaultYawMinRange, PLConstants.kDefaultYawMaxRange)
        mRollRange = PLRangeMake(PLConstants.kDefaultRollMinRange, PLConstants.kDefaultRollMaxRange)
        this.tempRange = PLRangeMake(0.0f, 0.0f)

        this.isZAxisEnabled = true
        this.isYAxisEnabled = this.isZAxisEnabled
        this.isXAxisEnabled = this.isYAxisEnabled
        this.isRollEnabled = true
        this.isYawEnabled = this.isRollEnabled
        this.isPitchEnabled = this.isYawEnabled

        this.isReverseRotation = false

        this.isYZAxisInverseRotation = true

        mPosition = PLPosition.PLPositionMake(0.0f, 0.0f, 0.0f)
        mRotation = PLRotation.PLRotationMake(0.0f, 0.0f, 0.0f)

        defaultAlpha = PLConstants.kDefaultAlpha
        alpha = defaultAlpha
    }

    override fun reset() {
        this.setRotation(0.0f, 0.0f, 0.0f)
        alpha = defaultAlpha
    }

    override var position: PLPosition
        get() = mPosition
        set(position) {
            run {
                this.x = position.x
                this.y = position.y
                this.z = position.z
            }
        }

    override fun setPosition(x: Float, y: Float, z: Float) {
        this.x = x
        this.y = y
        this.z = z
    }

    override var x: Float
        get() = mPosition.x
        set(x) {
            if (this.isXAxisEnabled) mPosition.x = PLMath.valueInRange(x, mXRange)
        }

    protected fun setInternalX(x: Float) {
        mPosition.x = PLMath.valueInRange(x, mXRange)
    }

    override var y: Float
        get() = mPosition.y
        set(y) {
            if (this.isYAxisEnabled) mPosition.y = PLMath.valueInRange(y, mYRange)
        }

    protected fun setInternalY(y: Float) {
        mPosition.y = PLMath.valueInRange(y, mYRange)
    }

    override var z: Float
        get() = mPosition.z
        set(z) {
            if (this.isZAxisEnabled) mPosition.z = PLMath.valueInRange(z, mZRange)
        }

    protected fun setInternalZ(z: Float) {
        mPosition.z = PLMath.valueInRange(z, mZRange)
    }

    override var rotation: PLRotation
        get() = mRotation
        set(rotation) {
            this.pitch = rotation.pitch
            this.yaw = rotation.yaw
            this.roll = rotation.roll
        }

    override fun setRotation(pitch: Float, yaw: Float) {
        this.pitch = pitch
        this.yaw = yaw
    }

    override fun setRotation(pitch: Float, yaw: Float, roll: Float) {
        this.pitch = pitch
        this.yaw = yaw
        this.roll = roll
    }

    override var pitch: Float
        get() = mRotation.pitch
        set(pitch) {
            if (this.isPitchEnabled) mRotation.pitch = this.getRotationAngleNormalized(pitch, mPitchRange!!)
        }

    protected fun setInternalPitch(pitch: Float) {
        mRotation.pitch = this.getRotationAngleNormalized(pitch, mPitchRange!!)
    }

    override var yaw: Float
        get() = mRotation.yaw
        set(yaw) {
            if (this.isYawEnabled) mRotation.yaw = this.getRotationAngleNormalized(yaw, mYawRange!!)
        }

    protected fun setInternalYaw(yaw: Float) {
        mRotation.yaw = this.getRotationAngleNormalized(yaw, mYawRange!!)
    }

    override var roll: Float
        get() = mRotation.roll
        set(roll) {
            if (this.isRollEnabled) mRotation.roll = this.getRotationAngleNormalized(roll, mRollRange!!)
        }

    protected fun setInternalRoll(roll: Float) {
        mRotation.roll = this.getRotationAngleNormalized(roll, mRollRange!!)
    }

    override var xRange: PLRange
        get() = mXRange!!
        set(xRange) {
            mXRange!!.setValues(xRange)
        }

    override fun setXRange(min: Float, max: Float) {
        mXRange!!.setValues(min, max)
    }

    override var xMin: Float
        get() = mXRange!!.min
        set(min) {
            mXRange!!.min = min
        }

    override var xMax: Float
        get() = mXRange!!.max
        set(max) {
            mXRange!!.max = max
        }

    override var yRange: PLRange
        get() = mYRange!!
        set(yRange) {
            mYRange!!.setValues(yRange)
        }

    override fun setYRange(min: Float, max: Float) {
        mYRange!!.setValues(min, max)
    }

    override var yMin: Float
        get() = mYRange!!.min
        set(min) {
            mYRange!!.min = min
        }

    override var yMax: Float
        get() = mYRange!!.max
        set(max) {
            mYRange!!.max = max
        }

    override var zRange: PLRange
        get() = mZRange!!
        set(zRange) {
            mZRange!!.setValues(zRange)
        }

    override fun setZRange(min: Float, max: Float) {
        mZRange!!.setValues(min, max)
    }

    override var zMin: Float
        get() = mZRange!!.min
        set(min) {
            mZRange!!.min = min
        }

    override var zMax: Float
        get() = mZRange!!.max
        set(max) {
            mZRange!!.max = max
        }

    override var pitchRange: PLRange
        get() = mPitchRange!!
        set(pitchRange) {
            mPitchRange!!.setValues(pitchRange)
        }

    override fun setPitchRange(min: Float, max: Float) {
        mPitchRange!!.setValues(min, max)
    }

    override var pitchMin: Float
        get() = mPitchRange!!.min
        set(min) {
            mPitchRange!!.min = min
        }

    override var pitchMax: Float
        get() = mPitchRange!!.max
        set(max) {
            mPitchRange!!.max = max
        }

    override var yawRange: PLRange
        get() = mYawRange!!
        set(yawRange) {
            mYawRange!!.setValues(yawRange)
        }

    override fun setYawRange(min: Float, max: Float) {
        mYawRange!!.setValues(min, max)
    }

    override var yawMin: Float
        get() = mYawRange!!.min
        set(min) {
            mYawRange!!.min = min
        }

    override var yawMax: Float
        get() = mYawRange!!.max
        set(max) {
            mYawRange!!.max = max
        }

    override var rollRange: PLRange
        get() = mRollRange!!
        set(rollRange) {
            mRollRange!!.setValues(rollRange)
        }

    override fun setRollRange(min: Float, max: Float) {
        mRollRange!!.setValues(min, max)
    }

    override var rollMin: Float
        get() = mRollRange!!.min
        set(min) {
            mRollRange!!.min = min
        }

    override var rollMax: Float
        get() = mRollRange!!.max
        set(max) {
            mRollRange!!.max = max
        }

    protected fun getRotationAngleNormalized(angle: Float, range: PLRange): Float {
        return PLMath.normalizeAngle(angle, tempRange!!.setValues(-range.max, -range.min))
    }

    override fun translate(position: PLPosition) {
        this.x = position.x
        this.y = position.y
        this.z = position.z
    }

    override fun translate(x: Float, y: Float) {
        this.x = x
        this.y = y
    }

    override fun translate(x: Float, y: Float, z: Float) {
        this.x = x
        this.y = y
        this.z = z
    }

    /**
     * rotate methods
     */
    override fun rotate(rotation: PLRotation) {
        this.pitch = rotation.pitch
        this.yaw = rotation.yaw
        this.roll = rotation.roll
    }

    override fun rotate(pitch: Float, yaw: Float) {
        this.pitch = pitch
        this.yaw = yaw
    }

    override fun rotate(pitch: Float, yaw: Float, roll: Float) {
        this.pitch = pitch
        this.yaw = yaw
        this.roll = roll
    }

    /**
     * clone methods
     */
    override fun clonePropertiesOf(`object`: PLIObject): Boolean {
        this.isXAxisEnabled = `object`.isXAxisEnabled
        this.isYAxisEnabled = `object`.isYAxisEnabled
        this.isZAxisEnabled = `object`.isZAxisEnabled

        this.isPitchEnabled = `object`.isPitchEnabled
        this.isYawEnabled = `object`.isYawEnabled
        this.isRollEnabled = `object`.isRollEnabled

        this.isReverseRotation = `object`.isReverseRotation

        this.isYZAxisInverseRotation = `object`.isYZAxisInverseRotation

        this.xRange = `object`.xRange
        this.yRange = `object`.yRange
        this.zRange = `object`.zRange

        this.pitchRange = `object`.pitchRange
        this.yawRange = `object`.yawRange
        this.rollRange = `object`.rollRange

        this.x = `object`.x
        this.y = `object`.y
        this.z = `object`.z

        this.pitch = `object`.pitch
        this.yaw = `object`.yaw
        this.roll = `object`.roll

        this.defaultAlpha = `object`.defaultAlpha
        this.alpha = `object`.alpha

        return true
    }

    /**
     * dealloc methods
     */
    @Throws(Throwable::class)
    protected open fun finalize() {
        mPosition = PLPosition()
        this.tempRange = null
        mZRange = null
        mYRange = null
        mXRange = null
        mRollRange = null
        mYawRange = null
        mPitchRange = null
    }
}